libyul/backends/wasm/TextTransform.cpp

```diff
@@ -162,12 +168,12 @@ string TextTransform::indented(string const& _in)
 string TextTransform::transform(wasm::FunctionDefinition const& _function)
 {
 	string ret = "(func $" + _function.name + "\n";
-	for (auto const& param: _function.parameterNames)
-		ret += "    (param $" + param + " i64)\n";
+	for (auto const& param: _function.parameters)
+		ret += "    (param $" + param.name + " " + param.type + ")\n";
 	if (_function.returns)
-		ret += "    (result i64)\n";
+		ret += "    (result " + _function.locals.at(0).type + ")\n";
```

Comment:
This looks weird, why is the 1st local the return variable?

Reply:
Because `FunctionDefinition` does not have a separate field describing the return value. It just has the `returns` flag and locals. Return values from the Yul function get translated to the first N locals and only after that the "real" locals are added. When the Wasm function returns, the first local gets assigned to the `result` variable and the remaining N-1 are assigned to globals.

It bothers me a bit too and I had to double check that it's really the way it works when I was changing it (wasn't obvious). Maybe it should be refactored too?

Reply:
Is there another way to do it?

Reply:
We could have a separate `returnVariable` field in `wasm::FunctionDefinition`.

Oh, I just realized that it wasn't there simply because in Wasm result does not have a name and there's not much more to say about it besides its presence or absence. So adding it was on me.

In that case I'll replace `bool returns` with `optional<wasm::Type> resultType`.

Reply:
Done.